### PR TITLE
WIP: Codeception

### PR DESCRIPTION
**Pull request by Fenikkusu**
This PR:

- [x] Adds Codeception Testing Framework
- [x] Covered by tests
- [ ] Doc PR: https://github.com/infection/site/pull/XXX

Fixes https://github.com/infection/infection/pull/457

Since Tobias is no longer maintaining the Codeception work he did, I've cleaned up my butchering of his work and am resubmitting it. As of now, the changes are working. However, I suspect there are a few pieces that may need to be altered due to a change in practice with the base system. I have not had the time to verify this. I am currently trying to get the changes working in the GitLab Runner, but am having some hard times with it (there is an error occurring, but no error output is given). A number of the changes I have added in are for the purpose of debugging the issues in the CI, but will likely be fine to maintain as additional features to the code.

I'm looking for any comments/thoughts on cleaning this up to get it fully implemented. On a side note, I'm aware a number of the checks will be invalid at this time. I've not fully cleaned up everything yet.

### CODE REVIEW

File: src/TestFramework/Factory.php [hunk 1]
```diff
@@ -116,10 +121,23 @@ public function create(string $adapterName, bool $skipCoverage): AbstractTestFra
             );
         }
 
+        if ($adapterName === TestFrameworkTypes::CODECEPTION) {
+            $codeceptionConfigPath = $this->configLocator->locate(TestFrameworkTypes::CODECEPTION);
+            $codeceptionConfigContent = file_get_contents($codeceptionConfigPath);
```
Comment (BackEndTea) [inline src/TestFramework/Factory.php]: Please add a `\assert(is_string($codeceptionConfigContent));`

File: src/Config/InfectionConfig.php [hunk 2]
```diff
@@ -167,4 +172,14 @@ public function getTestFramework(): string
     {
         return $this->config->testFramework ?? 'phpunit';
     }
+
+    public function getInitialTestsPhpOptions(): string
```
Comment (BackEndTea) [inline src/Config/InfectionConfig.php]: How about moving these new options to a separate PR?
It would reduce the size of this PR a little bit, and make it easier to review.

And could you also add tests for these new methods?

File: src/Command/InfectionCommand.php [hunk 3]
```diff
@@ -165,6 +166,13 @@ protected function configure(): void
                 InputOption::VALUE_NONE,
                 'Debug mode. Will not clean up Infection temporary folder.'
             )
+            ->addOption(
+                'error-timeout',
```
Comment (BackEndTea) [inline src/Command/InfectionCommand.php]: Do we really need to expose this option to our end users?I think it would lead to confusion, as its not relevant for most users.

Do you think we could chose a sensible default, and simply use that?

File: src/TestFramework/Codeception/Adapter/CodeceptionAdapter.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+        // "Warnings!" - e.g. when deprecated functions are used, but tests pass
+        $isWarning = (bool) preg_match('/^WARNINGS!/im', $output);
+
+        return $isOk || $isOkWithInfo || $isWarning;
+    }
+
+    public function getName(): string
+    {
+        return TestFrameworkTypes::CODECEPTION;
```
Comment (BackEndTea) [inline src/TestFramework/Codeception/Adapter/CodeceptionAdapter.php]: Please change this to `return 'Codeception';`. Its the way that Codeception is branded.

File: src/TestFramework/Codeception/CustomAutoloadFilePath.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Infection\TestFramework\Codeception;
+
+use Codeception\Events;
+use Codeception\Extension;
+
+class CustomAutoloadFilePath extends Extension
```
[the remark this answers is not in the document]
Comment (Fenikkusu) [reply src/TestFramework/Codeception/CustomAutoloadFilePath.php]: I thought I removed this file. With the bootstrap config setting, it no longer matters. Additionally, I couldn't find any usages anyways...

File: src/Command/InfectionCommand.php (hunk 3 above)
Comment (Fenikkusu) [reply src/Command/InfectionCommand.php]: Honestly, I'm not sure. In this particular instance (as it relates to codeception), it really only is a microsecond difference. My suggestion is leave it as a config file option and remove the cli arg, if you do want to remove it. That way, it isn't a mainstream option, but can still be configured.

File: src/TestFramework/Factory.php (hunk 1 above)
Comment (Fenikkusu) [reply src/TestFramework/Factory.php]: Have you any objections to a general clean up? I feel like the entire function can likely be refactored such that modifications to the file would not be necessary if a new framework were to be added. Not a really big issue, but something I was considering.

File: src/TestFramework/Codeception/CodeceptionExtraOptions.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
+/**
+ * Created by PhpStorm.
```
Comment (Fenikkusu) [inline src/TestFramework/Codeception/CodeceptionExtraOptions.php]: Need to update DocBlock

File: src/TestFramework/Factory.php (hunk 1 above)
Comment (BackEndTea) [reply src/TestFramework/Factory.php]: I'm all for general clean up. If you do so please move the clean up part to another PR. It allows us  to get the clean up part into the code base earlier, and makes reviewing this PR easier.

**Comment by charles-martin** [comment]
There isn't anything specific to be done (That I recall). More so that to make this merger easier, the other mergers need to be merged in. You are welcome to try the changes directly by specifying

```json
{
  "repositories" : [
     {
         "type": "github",
         "url": "https://github.com/twistersfury/infection"
     }
  ],
  "require-dev": {
    "infection/infection":"dev-codeception"
  }
}
```

in your composer.json (If you haven't already.)

**Comment by Fenikkusu** [comment]
Correction: I do need to still fix a few discussions...crud...

There isn't anything specific to be done (That I recall). More so that to make this merger easier, the other mergers need to be merged in. You are welcome to try the changes directly by specifying

```json
{
  "repositories" : [
     {
         "type": "github",
         "url": "https://github.com/twistersfury/infection"
     }
  ],
  "require-dev": {
    "infection/infection":"dev-codeception"
  }
}
```

in your composer.json (If you haven't already.)

**Comment by acrobat** [comment]
I've tried this feature in a codeception setup and everything seems to work. I just tested running the tests and getting results from infection, but this basic features seems to be working!

**Comment by edno** [comment]
Just tried the feature and I running Codeception I got a different outcome compared to PHPUnit :

### PHPUnit
11 mutants created and  **7 killed**
```shell
$ ./vendor/bin/infection --test-framework=phpunit
You are running Infection with xdebug enabled.
    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____ 
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/
 
Running initial test suite...

PHPUnit version: 7.5.6

   10 [============================] < 1 sec

Generate mutants...

Processing source code files: 2/2
Creating mutated files and processes: 11/11
.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out

....MMM..S.                                          (11 / 11)

11 mutations were generated:
       7 mutants were killed
       1 mutants were not covered by tests
       3 covered mutants were not detected
       0 errors were encountered
       0 time outs were encountered

Metrics:
         Mutation Score Indicator (MSI): 63%
         Mutation Code Coverage: 90%
         Covered Code MSI: 70%

Please note that some mutants will inevitably be harmless (i.e. false positives).
Dashboard report has not been sent: it is not a Travis CI

Time: 2s. Memory: 10.00MB
```

### Codeception
11 mutants created and **_10_ killed**
```shell
$ ./vendor/bin/infection --test-framework=codeception
You are running Infection with xdebug enabled.
    ____      ____          __  _
   /  _/___  / __/__  _____/ /_(_)___  ____ 
   / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
 _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
/___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/
 
Running initial test suite...

codeception version: 2.5.4

   18 [============================]  1 sec

Generate mutants...

Processing source code files: 2/2
Creating mutated files and processes: 11/11
.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out

.........S.                                          (11 / 11)

11 mutations were generated:
      10 mutants were killed
       1 mutants were not covered by tests
       0 covered mutants were not detected
       0 errors were encountered
       0 time outs were encountered

Metrics:
         Mutation Score Indicator (MSI): 90%
         Mutation Code Coverage: 90%
         Covered Code MSI: 100%

Please note that some mutants will inevitably be harmless (i.e. false positives).
Dashboard report has not been sent: it is not a Travis CI

Time: 3s. Memory: 10.00MB
```

Code used for assessing the feature : https://gist.github.com/edno/335b2246f026b269fa7e6aaaeb5c4707

**Comment by edno** [comment]
The generated file `codeception.initial.infection.yml` has incorrect paths. 
The relative path has been concatenated to the real path (excepted for `output` that use real path):
```yaml
paths:
    tests: ../../../../../../../../Users/edno/Workspace/demo-infection/tests
    output: /Users/edno/Workspace/demo-infection/tests/_support/infection/.
    data: ../../../../../../../../Users/edno/Workspace/demo-infection/tests/_data
    support: ../../../../../../../../Users/edno/Workspace/demo-infection/tests/_support
    envs: tests/_envs
actor_suffix: Tester
coverage:
    enabled: true
    include: ['../../../../../../../..//Users/edno/Workspace/demo-infection/src/*.php']
    exclude: {  }
extensions:
    enabled: [Codeception\Extension\RunFailed]
```

**Comment by Fenikkusu** [comment]
Unless I'm mistaken, the paths are correct. First note that you did get proper testing. Had these been wrong, it would of failed completely. If memory serves me, codeception bases its paths off the location of the config file. The config file is written to the temp directory for infection, so codeception paths off that. The relative paths are done such that codeception properly detects the paths.

**Comment by maks-rafalko** [comment]
Closing in favor of https://github.com/infection/infection/pull/800

**Comment by maks-rafalko** [comment]
Implemented in https://github.com/infection/infection/pull/800
